### PR TITLE
Np 2499 recognize query parameters

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: src/handler.js [hunk 1]
```diff
@@ -80,6 +80,13 @@ const createBadRequestDetails = (event) => {
...
+
 const problemsWithGetMethod = event => hasQueryParameters(event) ? createBadRequestDetails(event) : createNotFoundDetails(event)
 
-const createErrorDetails = (event) => isGetMethod(event) ? problemsWithGetMethod(event) : createMethodNotAllowedDetails(event)
+const hasValidQueryParameters = (event) => event.queryStringParameters.query !== undefined && hasValidParameterNames(event)
+
+const hasValidParameterNames = (event) => {
+  const parameterNames = ['query', 'year', 'start']
+  return Object.keys(event.queryStringParameters).filter(key => (!parameterNames.includes(key))).length === 0
+}
```
Comment (brinxmat) [inline src/handler.js:92]: While the code works perfectly, I wonder if we should view query as a standard case, e.g.
```
const hasValidQueryParameters = (event) => {
  const querySpec = [{ name: 'query', required: true }, { name: 'year', required: false }, { name: 'start', required: false }]
  const queryStringKeys = Object.keys(event.queryStringParameters);
  return isValidParameterName(event, queryStringKeys) && hasRequiredParameters(querySpec, queryStringKeys)
}

const isValidParameterName = (actualKeys, querySpec) => actualKeys.every(key => querySpec.map(param => param.name).includes(key))

const hasRequiredParameters = (actualKeys, querySpec) => {
  const required = querySpec.filter(param => param.required === true).map(param => param.name)
  return required.every(item => actualKeys.includes(item))
}
```
This can later be refactored to a generic config in an external class

**Review by brinxmat — CHANGES_REQUESTED** [review]
Can you at the same time inline the response on line 37?

File: src/handler.js [hunk 2]
```diff
@@ -62,7 +61,7 @@ const hasQueryParameters = (event) => {
   return 'queryStringParameters' in event && event.queryStringParameters !== undefined
 }
 
-const isValidRequest = (event) => routes.includes(event.path) && isGetMethod(event) && !hasQueryParameters(event)
+const isValidRequest = (event) => routes.includes(event.path) && isGetMethod(event) && (!hasQueryParameters(event) || hasValidQueryParameters(event))
```
Comment (axthosarouris) [inline src/handler.js:64]: line breaks and put the OR couple into a function

File: src/tests/unit/handler.test.js [hunk 3]
```diff
@@ -93,3 +93,37 @@ describe("Handler sets different 'Content-type' in respnse headers", () => {
...
+    const event = { path: '/journal', httpMethod: 'GET', queryStringParameters: queryStringParameters }
+    const response = await handler(event)
+    const emptyBody = '{}'
+    expect(response.statusCode).to.equal(httpStatus.OK)
+    expect(response.body).to.equal(emptyBody)
+  })
+})
+
+describe('Handler returns bad request when error in query ', () => {
+  it('response is 400 Bad Request when obligatory "query"-parameter is missing', async function () {
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:117]: general comment: It would be better if the naming was it('returns Bad Request when ...)

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: src/tests/unit/handler.test.js [hunk 4]
```diff
@@ -78,18 +78,52 @@ describe('Handler throws error when called with queryStringParameters', () => {
...
-  it("handler returns 'Content-Type' 'application/json' when responsecode is 200", async function () {
+  it("return 'Content-Type' 'application/json' when responsecode is 200", async function () {
     const event = { path: '/journal', httpMethod: 'GET' }
     const response = await handler(event)
     expect(response.statusCode).to.equal(httpStatus.OK)
     expect(response.headers).to.have.property('Content-Type')
     expect(response.headers['Content-Type']).to.equal('application/json')
   })
-  it("handler returns 'Content-Type' 'application/problem+json' when error occurs", async function () {
+  it("return 'Content-Type' 'application/problem+json' when error occurs", async function () {
```
Comment (axthosarouris) [inline src/tests/unit/handler.test.js:88]: "returns"

**Review by axthosarouris — APPROVED** [review]

File: src/handler.js [hunk 5]
```diff
@@ -58,11 +57,11 @@ const getProblemInstance = (event) => {
   return 'queryStringParameters' in event ? `${event.path}?${event.queryStringParameters}` : event.path
 }
 
-const hasQueryParameters = (event) => {
-  return 'queryStringParameters' in event && event.queryStringParameters !== undefined
-}
+const hasQueryParameters = (event) => 'queryStringParameters' in event && event.queryStringParameters !== undefined
```
Comment (brinxmat) [inline src/handler.js:60]: I wonder if we want:

```
const hasQueryParameters = (event) => Object.prototype.hasOwnProperty.call(event, 'queryStringParameters') && (event.queryStringParameters !== null || (Array.isArray(event.queryStringParameters) && event.queryStringParameters.length))

```

This has a couple of benefits, `x in y` interrogates properties of the object and inherited properties. In this case we are only interested in properties of the object. x !== undefined in practice does exactly the same thing. We should also test that the value is not an empty array because a non-array value is not OK. We also need to check that the value is not-null as we are negating the check

The test input needs to be updated for all these possibilities

**Review by brinxmat — CHANGES_REQUESTED** [review]

**Review by brinxmat — APPROVED** [review]
